Redesign

rvolosatovs opened the pull request:
Closes enarx/sallyport#17 

Docs at their current state are available at https://rvolosatovs.github.io/universe/rust/sallyport/index.html

This PR is the last one in the series of crate redesign, on top of which actual request handling (e.g. syscalls) will be implemented.

## Follow-ups (issues should be filed)
- Syscall implementations (feature-parity as first step) enarx/sallyport#34 https://github.com/enarx/sallyport/pull/20 
- Lint against `unsafe` enarx/enarx#1724 
- (tech debt) Macros for request definitions and possibly reducing boilerplate within the crate, especially for requests without allocations
- (tech debt) Improve documentation enarx/enarx#1725 enarx/sallyport#33 
- (tech debt) Profiling and following optimization of operations on the hot path.
- (tech debt) Use a designated directory for temporary test files, use `O_TMPFILE` where possible/feasible

rvolosatovs commented on src/v2/src/block.rs:
```diff
@@ -0,0 +1,165 @@
+// SPDX-License-Identifier: Apache-2.0
```
This whole module will probably live within `guest` module, i.e. `guest::Block`, still working out what makes most sense

Commit:
v2/guest/platform: Add module

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest/handler: Add module

Signed-off-by: Roman Volosatovs <roman@profian.com>

rvolosatovs commented:
We did an offline walkthrough of the PR with @npmccallum.
There will be a few bigger changes here, so if you are willing to look at this PR, it's most probably worth waiting for the next iteration to be ready before doing that.
Some of the changes that will be done in next iteration:
- Fix incorrect usages of `usize` vs `u8`
- Account for alignment on allocation
- `request::Syscall` will be removed, instead just plain `[usize; 7]` will be used
- `Platform.exit` will be removed and `_exit` syscall will be used instead.
- push functionality naming will be adjusted (with `reserved` and `copied` suffixes)
- raw pointers will be used in the block instead of references

haraldh commented:
I don't see any alignment promises/checks for the allocated buffer space. While this works for `&[u8]`, it might fail for more complex structures or `u64` after an uneven `[u8]`.

Also try implementing `recvfrom` to test your ergonomics.

haraldh commented:
> * Fix incorrect usages of `usize` vs `u8`
> 
> * Account for alignment on allocation
> 
> * `request::Syscall` will be removed, instead just plain `[usize; 7]` will be used
> 
> * `Platform.exit` will be removed and `_exit` syscall will be used instead.
> 
> * push functionality naming will be adjusted (with `reserved` and `copied` suffixes)
> 
> * raw pointers will be used in the block instead of references
> 
> * `guest::block::Addr` will be removed (no naked type aliases)

Oh.. I should have read that before the review :)

rvolosatovs commented:
NOTE: the implementation will be changed to rely on traits, which will roughly look like this:
```rust
pub trait Allocator {
    fn free(&self) -> usize;

    fn allocate_data_mut<T>(&mut self) -> Result<DataRefMut<T>>;
    fn allocate_data_slice_mut<T>(&mut self, len: usize) -> Result<DataRefMut<[T]>>;

    fn allocate_exchange(&mut self) -> Result<ExchangeRef>;

    fn allocate_data<T>(&mut self) -> Result<DataRef<T>> {
        self.allocate_data_mut().into()
    }

    fn allocate_data_slice<T>(&mut self, len: usize) -> Result<DataRef<[T]>> {
        self.allocate_data_slice_mut(len).into()
    }
}

pub trait Collect {
    type Item;

    // executed after `sally()`, which will write all data to destinations and return a reply
    fn collect(self) -> Self::Item;
}

pub trait Commit {
    type Item: Collect;

    // executed right before `sally()` after all requests are staged, writes data to the block
    fn commit(self) -> Self::Item;
}

pub trait Stage {
    type Item: Commit;

    fn stage<A: Allocator>(&self, alloc: &mut A) -> Result<Self::Item>;
}
```

rvolosatovs commented:
Really wanted to get the allocation in a good state this week, but it's time to have a weekend for me - I managed to port Read and Write to the new structure - pushed a (very) rough draft if you want to take a look. I will trim it down and clean a lot, did not get to that yet
CI warnings about unused variables make diff quite impossible to use, but you can see the code here https://github.com/enarx/sallyport/tree/448a7b184706667fb3ce223ab9a1235b5bba0dea/src/v2
the (low-level) allocation is done, on a higher level I will add necessary trait implementations per-type when implementing actual syscalls

Anyway, the idea is this:
Read: https://github.com/enarx/sallyport/blob/448a7b184706667fb3ce223ab9a1235b5bba0dea/src/v2/src/guest/syscall/read.rs#L12-L74
Write: https://github.com/enarx/sallyport/blob/448a7b184706667fb3ce223ab9a1235b5bba0dea/src/v2/src/guest/syscall/write.rs#L12-L70

These are consumed by the generic `sally` https://github.com/enarx/sallyport/blob/448a7b184706667fb3ce223ab9a1235b5bba0dea/src/v2/src/guest/handler.rs#L50-L71, which may be even exported across crate boundary eventually to allow for custom request implementations

The goal is to contain `unsafe{}` to `alloc` as much as possible and instead provide safe abstractions, which are very hard to use incorrectly.

Apart from that, we get write at-most-once semantics when attempting to sally a set of requests, and, with careful request implementations we can ensure that memory is written sequentially from the start of the block to the end, which may be nice for performance eventually (it's way too early to care about this now, but good to keep in mind). Note that I changed the order of reply-request pairs for that exact reason.
Block now looks like this: `[N, data(0..data_len), request(N), reply(N) request(N-1), reply(N-1) .., request(0), reply(0)]`
Given that everything is based on generic traits, everything is composable and we can easily hook into whichever part of the system we might want for things like smuggling requests, batching, tracing, logging etc.

rvolosatovs commented:
Both guest and host side needs to match 7 opaque registers to a concrete syscall to perform potential address translation. That means that there will be a huge `match` case on both sides and there's no way around that, unfortunately.

Original implementation relied on `Syscall` enum representing the registers as named, typed fields in a struct.
That has since been removed and the current approach essentially emulates that, except with a few drawbacks compared to the original enum struct-based approach.
- The whole match has to be duplicated on both sides, preferably exactly the same, otherwise everything below gets more and more painful.
- Over time inconsistencies may accumulate, which may result in bugs and complicate maintenance.
- It is easy to make a mistake, which is not easy to debug. E.g. consider that an implementation side accidentally types `(libc::SYS_recvfrom, [sockfd, buf, len, flags, addrlen, src_addr])` instead of `(libc::SYS_recvfrom, [sockfd, buf, len, flags, src_addr, addrlen])`. The mistake is non-obvious. In case `recvfrom` is misbehaving, there's now an additional huge match statement that has to be checked. Struct enum approach fixes that, because it gives us a single source of truth with named, strongly-typed arguments, which completely eliminates this class of bugs (assuming correct implementation of translation, of course, but that is pretty trivial and correctness can be assured by unit tests, which is not true for number matching approach)
- Now that there are 2 sides where syscalls have to be handled, it is easy to forget to implement a syscall on a side. Enum structs fix that, by taking the burden of checking this from shoulders of the developer to the compiler, which means that a missing syscall is now a compile-time error, not a runtime one.

 Given the reasoning above, I will revive the enum struct approach as was used here, albeit only with syscalls actually implemented defined:
 https://github.com/enarx/sallyport/blob/fb75fb935843ea8238500ed855c184c2f85a20ff/src/v2/src/request/syscall.rs#L10-L24

haraldh commented:
I guess we can generate `TryFrom<[usize; 7]> for Syscall<'a>` and `From<Syscall<'a>> for Segment` with a derive proc_macro or maybe even with a simple macro_rules macro.

rvolosatovs commented:
> I guess we can generate `TryFrom<[usize; 7]> for Syscall<'a>` and `From<Syscall<'a>> for Segment` with a derive proc_macro or maybe even with a simple macro_rules macro.

Yes, I would like to explore how we can utilize macros to reduce boilerplate here and, especially, in the request implementations. I don't think that should be a requirement for this PR though.

rvolosatovs commented on src/v2/src/host/address_space.rs:
```diff
@@ -0,0 +1,63 @@
+        let start = (self.0.start as usize).checked_add(offset).ok_or(EINVAL)?;
+        let end_ptr = start.checked_add(size).ok_or(EINVAL)? as _;
+        let start_ptr = start as _;
+        if !self.0.contains(&start_ptr) || !self.0.contains(&end_ptr) {
```
This might be off-by-one, I have to check the docs, which I will do shortly

rvolosatovs replied:
FTR, ixed in https://github.com/enarx/sallyport/pull/18/commits/e2d97b00e22b58b2328aa46fe4cac46734d211dc

rvolosatovs commented on src/v2/src/guest/handler.rs:
```diff
@@ -0,0 +1,87 @@
+        if count > 0 {
+            self.platform.sally()?;
+
+            if unsafe { self.count.read() } != count {
```
Is this sufficient?
Do we want to e.g. compute a checksum/hash of committed data and verify integrity here?

npmccallum commented:
> * (tech debt) Use a designated directory for temporary test files, use `O_TMPFILE` where possible/feasible

We should just `EINVAL` or similar if we get `O_TMPFILE`. It is not supported.

npmccallum commented on src/v2/src/request/mod.rs:
```diff
@@ -0,0 +1,47 @@
+
+    fn try_from(v: usize) -> Result<Self> {
+        match v {
+            v if v == Kind::Syscall as usize => Ok(Kind::Syscall),
```
If you're going to use the `Kind` discriminators this way, they should be defined. So:

```
enum Kind {
  Syscall = 0,
}
```

rvolosatovs replied:
It's 0 by default according to the Rust reference https://doc.rust-lang.org/reference/items/enumerations.html#custom-discriminant-values-for-fieldless-enumerations

> If there is no data attached to any of the variants of an enumeration, then the discriminant can be directly chosen and accessed.
>
>These enumerations can be cast to integer types with the as operator by a numeric cast. The enumeration can optionally specify which integer each discriminant gets by following the variant name with = followed by a constant expression. If the first variant in the declaration is unspecified, then it is set to zero. For every other unspecified discriminant, it is set to one higher than the previous variant in the declaration.

I do think it's best to set a value explicitly though to ensure it's preserved in the future and does not get accidentally modified.

rvolosatovs commented on src/v2/src/guest/syscall/result/linux/x86_64.rs:
```diff
@@ -0,0 +1,78 @@
+
+impl<T> Result<T> {
+    #[inline]
+    pub const unsafe fn errno_unchecked(errno: c_int) -> Self {
```
This is necessary, because const options are unstable

Commit:
v2/guest/alloc: Promote phase transitions to public trait

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest: Fix {In,Out}Ref naming in documentation

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest/alloc: Add block assertion test helper

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest/alloc: Use type divider for numeric test constants

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest/alloc: Add support for `reserve_*` operations

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/{guest,host}: Add `read` syscall

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/guest/alloc: Ensure syscall sizes are multiples of `size_of::<usize>`

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item: Implement IntoIterator for Block

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/item: Skip unknown items in the iterator

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2/host: Store syscall rets as array of references

This better matches the usage pattern and better aligns with safer
approach.

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2: Add integration test

Signed-off-by: Roman Volosatovs <roman@profian.com>

Commit:
v2: Enable `asm` by default

miri does not support `asm` currently, therefore disable default
features for CI test invocation

Signed-off-by: Roman Volosatovs <roman@profian.com>
